Add tooltips on mult inputs

juzen2003 opened the pull request:
- Fixes #972. Fixes #425.
- This is the continuation of https://github.com/SETI/pds-opus/pull/1179
- Use the master as the base and merge the old tooltips branch.

- Current Implementation:
    - Create mult_tooltips.json to store tooltips (HTML markup are allowed in the content) for mult options.
    - Add mult tooltips context name to contexts.csv
    - Store mult tooltips to dictionary (definitions table)
    - In definitions table, the term will be the same as the mult option value in the mult tables, and context will be ```MULT_{slug.upper()}```. And the definitions (tooltips) will be fetched by ```get_def_for_tooltip later```.
    - Since tooltips are stored in definitions table, we can update tooltips by ```python main_opus_import.py --import-dictionary``` instead of re-import the whole database.

Note: for now testing toolstips are added for the options in mission widget. We can see the tooltips with different HTML markup when hovering over Galileo, New Horizons, and Voyager

juzen2003 committed:
Setup tooltipster and put tooltips at the last column of mult tables.
Currently only MultObsGeneralMissionId has tooltips instantiate by tooltipster.

juzen2003 committed:
Pass current selections to widget template for customized mult inputs.
- In widget.html (line 66-70), check the selected mult inputs if current
  selections is not empty.
- In views.py (line 485-490), pass current selections to widget template.
- In widgets.js (line 1141), removed unused functions when initializing
  tooltipster.

juzen2003 committed:
Renamed mult table "tooltips" field to "tooltip", and also change field
type to "text".

juzen2003 committed:
Add the functionality to use customized widget input html for mult group.
- Create customized_input.html for customized mult input. This can be
  re-used for customized mult or mult group inputs.
- In widget.html (line 61-81), call customized_input.html for both customized
  inputs for mult and mult group.
- In views.py (line 421-449), get the options for customized mult inputs
  if we decided to use the customized html for them.

juzen2003 committed:
- Convert default tooltips in cart.html to use tooltipster.
- Rename cust-tooltip-* to op-mult-tooltip-*
- Use op-cart-tooltip for customized tooltips in cart.html
- Add constants tooltips_max_width and tooltips_theme to set width and
  theme for customized tooltips

juzen2003 committed:
Convert default tooltips to use tooltipster in detail tab.
- Use .op-detail-tooltip for customized tooltips in detail.html
- Use .op-detail-metadata-tooltip for customized tooltips in metadata
  area of the detail tab.

juzen2003 committed:
Convert default tooltips to use tooltipster in base.html.

juzen2003 committed:
Convert default tooltips to use tooltipster in menu.html.

juzen2003 committed:
Convert default tooltips to use tooltipster the followings:
- Image in the detail tab
- Search widget header
- Clean button in download popover

juzen2003 committed:
Convert default tooltips to use tooltipster in browse tab, including
gallery/table views, and metadatabox.

juzen2003 committed:
Convert default tooltips to use tooltipster in the followings:
- Add/remove input icons in the widgets
- Buttons in the sorting pills in the navbar
- (i) in select metadata menu

juzen2003 committed:
Remove unwanted console.log.

juzen2003 committed:
Add tooltip to all MultObs* models, and add the functionality to get the
tooltip for each mult optoin from get_mult_tooltip.
- In model.py, add tooltip to all MultObs* models.
- In views.py (line 427-435), replace invalid characters in the mult
  options with '_'. This will make sure we don't assign invalid
  characters to HTML class/id for customized tooltips.
- In customized_input.html (line 13, 19), make sure we use string (tp_id)
  without invalid characters for html class/id for tooltipster.
- In settings.py (line 377), create INVALID_CLASS_CHAR to store invalid
  characters for HTML class/id.
- In config_data.py (line 1105-1115), create a dictionary to store
  tooltip for mult options.
- In do_import.py (1442-1445), make sure we get the mult tooltip from
  populate functions and store it in database.
- In import_util.py (line 246-257), create get_mult_tooltip to get the
  mult options tooltips from TOOLTIPS_FOR_MULT.
- In populate_obs_instrument_COISS.py (line 652-657), add the ability
  to add mult tooltips for COISS filter.

juzen2003 committed:
Add the ability to have mult tooltips for coiss_target_desc.
- In populate_obs_instrument_coiss_target_desc (line 699-705), check
  if tooltips for mult_obs_instrument_coiss_target_desc is added.

juzen2003 committed:
Add the ability to implement customized mult tooltips for all dynamically
generated mult widgets.

juzen2003 committed:
Fix django tests and removed unused logic in helper_voyager_planet_id.
- In populate_obs_mission_voyager.py (line 50), removed unused logic in
  helper_voyager_planet_id.

juzen2003 committed:
Update the tooltip theme and behaviors in browse table view.
- In browse.js (line 1587-1608), add the tooltip behaviors in browse table
  view.
- In browse.js (line 71-85), reposition table view tooltip to make sure
  tooltip pops out right next to the cursor when the cursor is moving
  around the same row.

juzen2003 committed:
Update the selector to make sure tooltips in both browse/cart data table
work correctly.
- In browse.js (line 74), update the selector to make sure tooltips behviors
  are correct in both browse/cart table view.

juzen2003 committed:
Clean up code styles found by JShint

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 972_425_tooltips_on_mults

juzen2003 committed:
Clean up code styles in opus.js (change '' to "")

juzen2003 committed:
Fixed the tooltip for "View Gallery/Table" & "Sort by".
- In browse.js (line 1398-1400, 1409-1411), after tooltipster is initialized,
  if we want to change the title, we need to do it by changing the content
  instead of updating the title attribute.
- In sortMetadata.js (line 22-29), initialize tooltips for "Sort by"
  using tooltipster

juzen2003 committed:
Convert default tooltip of data table header to use tooltipster.
-In browse.js (line 1694), add a tooltipster class for data table header
 and initialize it later.

juzen2003 committed:
Make sure new lines exists by replacing "\r" and "\n" in title attribute
with "<br>" for tooltipster instances.

juzen2003 committed:
Merge branch '972_425_tooltips_on_mults' into 972_425_tooltips_on_mults_new

juzen2003 committed:
- Create mult_tooltips.json to store tooltips for mult options.
- Add mission mult tooltips context name to contexts.csv
- Store mission mult tooltips to dictionary (definitions table)

juzen2003 committed:
In definitions table, set the term to be the same as the mult option
value in the mult tables, and set the context to be MULT_{slug.upper()}.
The definitions (tooltips) will be fetched by get_def_for_tooltip later.

juzen2003 committed:
Use safe filter for the mult tooltips variable inserted into Django template.
This allows us to add HTML markup to the tooltip content.

juzen2003 committed:
Removed commented out code.

rfrenchseti commented on opus/import/config_data.py:
```diff
@@ -152,3 +152,21 @@
     63: 'Madrid 70m',
     65: 'Madrid 34m'
 }
+
```
This can go away, right?

rfrenchseti commented on opus/import/import_util.py:
```diff
@@ -243,6 +243,17 @@ def safe_column(row, column_name, idx=None):
         return None
     return row[column_name][idx]
 
+def get_mult_tooltip(table_name, field_name, option_val):
```
This can go away, right?

rfrenchseti commented on opus/import/do_dictionary.py:
```diff
@@ -31,6 +31,10 @@ def create_import_definitions_table():
     json_def_path = (opus_secrets.DICTIONARY_JSON_SCHEMA_PATH +
                      '/internal_def*.json')
     json_list += glob.glob(json_def_path)
+    # Tooltips for mult 
```
for *mults*

rfrenchseti commented on opus/import/do_dictionary.py:
```diff
@@ -31,6 +31,10 @@ def create_import_definitions_table():
                      '/internal_def*.json')
     json_list += glob.glob(json_def_path)
+    # Tooltips for mult 
+    json_tooltips_def_path = (opus_secrets.DICTIONARY_JSON_SCHEMA_PATH +
```
I think we want to do the same thing here that's done for obs*.json above . Allow multiple mult_tooltips files like mult_tooltips_mission and mult_tooltips_instrument. Otherwise the file is going to get enormous, and this allows multiple people to work on different widgets at the same time. Since you're already using glob this just means changing the filename to mult_toolips*.json.

rfrenchseti commented on opus/import/table_schemas/mult_template.json:
```diff
@@ -37,6 +37,11 @@
         "field_type": "char100",
         "field_default": null
     },
+    {
```
This can go away, right? We don't store the tooltip in the mult table anymore.

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1204,17 +1205,17 @@ def import_observation_table(instrument_obj,
             if field_type.startswith('flag'):
                 if column_val in [0, 'n', 'N', 'no', 'No', 'NO', 'off', 'OFF']:
                     if field_type == 'flag_onoff':
-                        column_val = 'Off'
```
What changed to require this? Isn't mult_label set further down below based on colum_val.title()?

rfrenchseti commented on opus/application/apps/dictionary/views.py:
```diff
@@ -13,8 +13,10 @@ def get_def_for_tooltip(term, context):
-        log.error('No tooltip definition for context "%s" term "%s"',
-                  context, term)
+        # We allow mult tooltips to be None
+        if not context.startswith('MULT'):
```
Change to MULT_ just to be cautious

rfrenchseti commented on dictionary/contexts.csv:
```diff
@@ -21,3 +21,4 @@ NEWHORIZONS,New Horizons Mission,OPUS_GENERAL
 NHLORRI,New Horizons Long-Range Reconnaissance Imager (LORRI),NEWHORIZONS
 NHMVIC,New Horizons Multispectral Visible Imaging Camera (MVIC),NEWHORIZONS
 OPUS_PRODUCT_TYPE,OPUS Product Types,NULL
+MULT_MISSION,OPUS Mission Ids,NULL
```
This isn't needed for now since we never actually use the contexts table. Eventually we'll want to add contexts automatically as the mult tooltips are loaded.

juzen2003 replied:
There is a constraint that requires the context field in definitions to have a foreign key to contexts table name field. If we remove the entry in context.csv, we will get the following error when running import.
```
Failed to insert row into "imp_definitions": Cannot add or update a child row: a foreign key constraint fails (`opus3_test_group_inst_id`.`imp_definitions`, CONSTRAINT `imp_definitions_ibfk_1` FOREIGN KEY (`context`) REFERENCES `imp_contexts` (`name`) ON UPDATE CASCADE)
```

rfrenchseti replied:
OK, but then this isn't the way to do it because you'd have to add a new context to this file every time you wanted to add tooltips to a new mult. That is an unnecessary dependency. Instead, import_dictionary should create the contexts for each mult context as it's needed.

rfrenchseti commented on opus/application/apps/ui/templates/ui/customized_input.html:
```diff
@@ -0,0 +1,28 @@
+                    {% else %}
+                        <input type="checkbox" name="{{ slug }}" value="{{ choice }}" class="multichoice" id="{{ slug }}_{{ id }}">
+                    {% endif %}
+                <!-- Assign an unique html structure to data-tooltip-content for each different individual tooltip-->
```
a unique

rfrenchseti commented on opus/application/apps/ui/templates/ui/customized_input.html:
```diff
@@ -0,0 +1,28 @@
+                    <div class="tooltip-templates">
+                        <span id="tooltip-content-{{ slug }}-{{ tp_id }}">
+                            <strong>{{ tooltip|safe }}</strong>
+                            <!-- <a class="op-mult-tooltip-link" href="https://pds-rings.seti.org/" target="_blank"><strong>{{ tooltip }}</strong></a> -->
```
Remove comment

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -283,6 +284,14 @@ def _update_form_with_grouping(form, form_vals, glabel, gvalue):
     auto_id = True
     selections = {}
     extras = {}
+    customized_input = slug in settings.MULT_WIDGETS_WITH_TOOLTIPS
```
We shouldn't need MULT_WIDGETS_WITH_TOOLTIPS anymore. Instead look through the mult values and see if any of them have tooltips. If they do, then you need a customized input. If they're all None, then you don't need one.

rfrenchseti commented on opus/application/settings.py:
```diff
@@ -373,3 +373,8 @@
 DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'
+
+# Mult widgets that required tooltips
+MULT_WIDGETS_WITH_TOOLTIPS = ['mission']
```
Get rid of

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -65,6 +70,22 @@ var o_browse = {
+        // Also when moving around in the same table row, reposition the tooltip so that
+        // it stays right next to the cursor. Work for both browse/cart data table.
+        $(".op-data-table tbody").on("mousemove", function(e) {
+            // We will wait for the stop of mouse moving, and then show the tooltip
```
for mouse moving to stop

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1858,6 +1883,46 @@ var o_browse = {
+            debug: false,
+            contentAsHTML: true,
+            functionBefore: function(instance, helper){
+                // Make sure all other tooltips is closed before a new one is open
```
are closed. is opened.

rfrenchseti commented on opus/application/static_media/js/browse.js:
```diff
@@ -1858,6 +1883,46 @@ var o_browse = {
+                let tooltipWidth = position.size.width;
+                let offsetToRightWindow = 5;
+                let windowWidth = helper.geo.window.size.width;
+                // When the cursor is very close to the right edget of the window, we have
```
right edge.

rfrenchseti commented on opus/application/static_media/js/opus.js:
```diff
@@ -151,6 +151,11 @@ var opus = {
+    // Tooltipster settings
+    tooltips_max_width: 300,
+    tooltips_theme: "tooltipster-light",
+    tooltips_delay: 1500,
```
I think this is pretty slow. Normal tooltips pop up a lot faster. Like maybe a half second? Is there any way to have the (i) info buttons show up fast and the mult values take longer?

juzen2003 replied:
we can assign different delays for different tooltips. I modified it to have 1.5 second delay for mult tooltips and 0.5 second delay for the rest of tooltips.

rfrenchseti commented on opus/application/static_media/js/widgets.js:
```diff
@@ -437,6 +437,10 @@ var o_widgets = {
             } else {
                 lastExistingSetOfInputs.append(orLabel);
                 cloneInputs.find(".op-or-labels").remove();
+                // remove the cloned trash icon, and attach a new so that the tooltipster
```
a new <what> so...

juzen2003 committed:
Removed unused variable (TOOLTIPS_FOR_MULT) and function (get_mult_tooltip)
in config_data.py & import_util.py

juzen2003 committed:
Remove tooltip field in mult_template.json since it's not used anymore.

juzen2003 committed:
Remove MULT_WIDGETS_WITH_TOOLTIPS and change the intelligence of setting
customized_input.

juzen2003 committed:
Use a slower tooltips pop out delay for mult tooltips, and clean up
comments based on github feedback.

juzen2003 committed:
Rename mult_tooltips.json to mult_tooltips_mission.json

juzen2003 committed:
Update the step to store mult tooltips entry to contexts table before
storing it into definitions table.
- In do_dictionary.py (line 98-109), add mult tooltips entry into contexts
  table before storing it into definitions table. We will leave contexts.csv
  untouched.

juzen2003 committed:
- Set max width of the tooltips window to 500px
- Reduce the font size of the tooltips contents to 12px
- Set the font color of the tooltips contents to black

juzen2003 committed:
Increate tooltips font size to 13px.

dstopp commented on opus/application/static_media/css/opus.css:
```diff
@@ -3475,3 +3475,23 @@ when screen is narrow (cart tab) */
+
+/* set font size and color of the tooltips contents */
+.tooltipster-content {
+    font-size: 13px;
```
I think the font size should be responsive; looks like it stays at 13px even when the screen size is small.

dstopp commented on opus/application/static_media/js/browse.js:
```diff
@@ -1878,7 +1943,7 @@ var o_browse = {
         // let addallIcon = "<button type='button' data-toggle='modal' data-target='#op-addall-to-cart-modal' " +
         let addallIcon = "<button type='button'" +
-                         "class='op-table-header-addall btn btn-link'" +
+                         "class='op-table-header-addall btn btn-link op-addall-icon-tooltip'" +
```
Please consider renaming "op-addall-icon-tooltip" to "op-icon-tooltip-addall" for class naming consistency

dstopp commented on opus/application/static_media/js/menu.js:
```diff
@@ -68,6 +68,13 @@ var o_menu = {
+                maxWidth: opus.tooltips_max_width,
+                theme: opus.tooltips_theme,
+                delay: opus.tooltips_delay,
+                debug: false,
```
do the other tipsters need "debug: false" as well, or can this be removed?

dstopp commented on opus/application/static_media/js/opus.js:
```diff
@@ -1495,4 +1509,11 @@ $(document).ready(function() {
+        maxWidth: opus.tooltips_max_width,
+        theme: opus.tooltips_theme,
+        delay: opus.tooltips_delay,
+        debug: false,
```
same question; does "debug: false" need to be here or in any of the others?

dstopp commented on opus/application/static_media/js/widgets.js:
```diff
@@ -1442,6 +1461,21 @@ var o_widgets = {
                 opus.updateOPUSLastSelectionsWithOPUSSelections();
             }
+
+            // TESTING
```
is this still test code?

juzen2003 committed:
- Renamed op-addall-icon-tooltip to op-icon-tooltip-addall
- Shrink the tooltips font size to 90% (original size: 13px) when screen
is small
- Remove all debug: false in tooltipster initialization

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 972_425_tooltips_on_mults_new

juzen2003 committed:
Rename op-metadatabox-tooltip to op-metadatabox-edit-tooltip in metadataboxHtml.

juzen2003 committed:
Fixed the issue of missing metadata preview image by changing selector
from #galleryViewContents to .op-metadata-detail-view-body in
updateMetadataDetailView

juzen2003 committed:
Add tooltips for "x", maximize and minimize icons in metadatabox

juzen2003 committed:
Removed unsued variable in opus.js

juzen2003 committed:
Update comments on functionPosition of .op-browse-table-tooltip and
.op-browse-gallery-tooltip

juzen2003 committed:
- Reduce tooltip max width to 350px
- Update styling of tooltips (op-custom-tooltip-theme) to have black
  font on white background
- Make sure the cursor on (i) of search widgets is the default style.

juzen2003 committed:
Make sure all tooltips are using tooltipster.
- Add class op-detail-tooltip to op-metadata-detail-add element
- Add class op-detail-tooltip to op-metadata-detail-remove element
- Add class op-detail-tooltip to op-detail-share element
- Add class op-input-action-tooltip to input delete buttons
- Add class op-detail-cart-tooltip to op-detail-cart element
-  class op-browse-gallery-tooltip to op-last-modal-overlay element
- Add class op-browse-gallery-tooltip to op-detail-overlay element

juzen2003 committed:
Move class op-detail-tooltip from op-detail-share to the a tag element
under op-detail-share

juzen2003 committed:
Fixed the remove input icon by removing duplicated op-input-action-tooltip
class assigned to the icon.

juzen2003 committed:
- Reduce tooltip max width to 350px.
- Update styling of tooltips (op-custom-tooltip-theme) to have black font
  on white background.
- Make sure the cursor on (i) of search widgets is the default style.
- Make sure all tooltips are converted to use tooltipster.
- Make sure when hovering over metadatabox preview image, detail preview
  image, and table thumbnail image, the tooltips will follow the cursor
  instead of staying on the top edge of the element.

juzen2003 committed:
Move the tooltipster initialization (for .op-browse-gallery-tooltip &
.op-browse-table-tooltip) step from renderGalleryAndTable to loadData.
This will avoid initializing tooltipster on those elements twice when
we attached the ps event listeners later.

juzen2003 committed:
Remove the mission tooltips definitions in mult_tooltips_mission.json

juzen2003 committed:
Set max width for tooltips in download pane to 350px, and max width for
rest of tooltips to 450px.

juzen2003 committed:
- Set the max width of the tooltip to 680px
- Rename onImgMouseMoveHandler to onMouseMoveHandler

juzen2003 committed:
Change padding around the tooltips text to 3px 5px.

juzen2003 committed:
Reduce the padding for tooltips contents, and also use hyphens: auto; to
have "-" when a word is breaking into two lines.

juzen2003 committed:
Update the padding on tooltips contents, make sure the left & right paddings
are the same. Also make sure the tooltip for the first pill of sort by
is split into two lines.

rfrenchseti reviewed (changes requested):
When you click on the "cart add" or "cart trash" icon in either the gallery view or the slideshow, OPUS tries to change the tooltip text but ends up adding an old-style tooltip so there are now two tooltips on that icon, one saying "Add" and one saying "Remove".

rfrenchseti commented on opus/import/table_schemas/mult_tooltips_mission.json:
```diff
@@ -0,0 +1,32 @@
+[
```
This file should just be removed entirely.

juzen2003 committed:
Make "Sort by" in browse and cart tabs have two-line tooltips.

juzen2003 committed:
Remove mult_tooltips_mission.json.

juzen2003 committed:
Update all tooltips contents changes by calling .tooltipster("content", new_tooltip)
instead of .attr("title", new_tooltip) or .prop("title", new_tooltip)

juzen2003 committed:
In browse/cart tabs metadata slide view, put back "(or use spacebar)"
to the tooltips of the add/remove icons.

dstopp commented on opus/application/apps/cart/templates/cart/cart.html:
```diff
@@ -20,16 +20,16 @@
                 </li>
                 <li class="list-inline-item">
-                    <a href="#" class="op-download-csv btn btn-outline-primary" download title="Download a CSV of selected metadata for all observations in the cart"><i class="fas fa-file-csv"></i>&nbsp;Metadata CSV</a>
+                    <a href="#" class="op-download-csv btn btn-outline-primary op-cart-tooltip" download title="Download a CSV of selected metadata for all observations in the cart"><i class="fas fa-file-csv"></i>&nbsp;Metadata CSV</a>
```
I don't believe "download" belongs (by itself) in the html part of this <a> tag

rfrenchseti replied:
Actually it does. See, for example: https://www.w3schools.com/TAGS/att_a_download.asp

dstopp commented on opus/application/settings.py:
```diff
@@ -373,3 +373,6 @@
 DEFAULT_AUTO_FIELD = 'django.db.models.AutoField'
+
+# We don't want to have these characters in HTML class or ID
+INVALID_CLASS_CHAR = '~!@$%^&*()+=,./;:"?><[]\{}|`# '
```
Perhaps a more detailed explanation as to why we care about this, ie., "for tooltips" or something of that nature.

dstopp commented on opus/application/static_media/css/opus.css:
```diff
@@ -489,6 +489,10 @@ ul.op-search-menu {
 }
 
+.widget .fa-info-circle:hover {
+    cursor: default;
```
I think all the .fa-info-circle:hover <and perhaps other tooltip locations> should have the same cursor

dstopp commented on opus/application/static_media/js/opus.js:
```diff
@@ -151,6 +151,19 @@ var opus = {
+    multTooltipsDelay: 1500,
+
+    // the x & y coordinate of the current cursor when moving mouse
+    mouseX: 0,
```
is this specifically for tooltips or general purpose?

juzen2003 replied:
This is tracking the mouse position, and that mouse position will be used to set the proper position for the tooltips. It can be used to other places where we need to track the position of the mouse, but for now, it's only used for setting the the position of the tooltips.

dstopp replied:
perhaps it would be useful to add just a bit more detail per your comment above.

juzen2003 committed:
- Fixed the wrong class name "ob-base-tooltip", change to "op-base-tooltip"
- Make sure black info circles have the default cursor style.
- Update comment for INVALID_CLASS_CHAR

juzen2003 committed:
Update the comment for opus.mouseX and opus.mouseY

juzen2003 committed:
- Add some distances between the tooltips and left & right edges of the
browser window.
- Increase the left & right spaces to 5px between the tooltips text and
tooltips box.

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 972_425_tooltips_on_mults_new

juzen2003 committed:
Use CDN for tooltipster

juzen2003 committed:
Removed unused CDN fallback tooltipster files.
